docs/design/design.md

```diff
@@ -135,56 +192,82 @@ Other flags that are set:
 
 Other flags that are set unconditionally:
  - The `BootstrapSigner` and `TokenCleaner` controllers are enabled
  - `--root-ca-file` to `ca.crt`
- - `--cluster-signing-cert-file` to `ca.crt`
- - `--cluster-signing-key-file` to `ca.key`
+ - `--cluster-signing-cert-file` to `ca.crt`, expect  if `kubeadm` is running in "ExternalCA"
+ - `--cluster-signing-key-file` to `ca.key`, expect  if `kubeadm` is running in "ExternalCA"
  - `--service-account-private-key-file` to `sa.key`
  - `--use-service-account-credentials` to `true`
 
+Please note that:
+
+1. if `kubeadm` is running in "ExternalCA" mode, `--cluster-signing-cert-file` and  `--cluster-signing-key-file` will be set to  `""` .
+
 #### Scheduler
 
 kubeadm doesn't set any special scheduler flags.
 
-Common properties for the control plane components:
- - Leader election is enabled for both the controller-manager and the scheduler
- - `HostNetwork: true` is present on all static pods since there is no network configured yet
+### Generate Static Pod Manifest for local etcd
+
+If the user specified an external etcd this step will be skipped, otherwise a static manifest file will be generated for creating a local etcd instance running in a pod with following attributes:
 
-#### Wait for the control plane to come up
+- listen on `localhost:2379` and use `HostNetwork=true`
+- have `PodSpec.SecurityContext.SELinuxOptions.Type=spc_t` set because of https://github.com/kubernetes/kubeadm/issues/107
+- make a `hostPath` mount out from the `dataDir` to the host's filesystem
+- Any extra flags specified by the user
+
+Please note that:
+
+1. The image, for version  `3.0.17` /current architecture, will be pulled from `gcr.io/google_containers`. In case an alternative image repository is specified this one will be used; In case an alternative image name is specified, this one will be used.
+2. in case of `kubeadm`  executed in the `--dry-run` mode, static pods files are written in a temporary folder.
+
+### Wait for the control plane to come up
 
 This is a critical moment in time for kubeadm clusters.
 kubeadm waits until `localhost:6443/healthz` returns `ok`
 
 kubeadm relies on the kubelet to pull the control plane images and run them properly as Static Pods.
 But there are (as we've seen) a lot of things that can go wrong. Most of them are network/resolv.conf/proxy related.
 
-### Phase 4+: Post-bootstrap Phases
+After the control plane is up, kubeadm completes a couple of tasks described in following paragraphs.
+
+### Saves MasterConfiguration in a ConfigMap for later reference
+
+kubeadm saves the MasterConfiguration, that contains the flags/parameters passed to `kubeadm init`,  in a ConfigMap named `kubeadm-config ` under `kubernetes-system` namespace.
 
-kubeadm completes a couple of tasks also after the control plane is up, namely these tasks:
+This will ensure that kubeadm action executed in future (e.g `kubeadm upgrade`) will get the same flags/parameters used for creating the cluster.
```

Comment:
kubeadm _actions_

Will be able to determine the actual/current cluster state and make new decisions based on that data